### DISCUSSION TITLE
Play weather report from an HTTP server

### DISCUSSION THREAD

**Discussion by Gvolten** [AzuraCast/AzuraCast#7321] [Q&A] [answered]
Hi. I have a custom-coded program that creates a simple weather report via TTS and hosts it on an HTTP server as an mp3 file. I'd like to play this report every hour on my station, but I cannot figure out how to set up the Remote URL playlist like this. It does not show the options for playing at a certain time. Does anyone know how to do this? Thanks.

**Comment by Mozitux** [comment]
Hi, is your tts server on the same as your Azuracast instance?

**Reply by Gvolten** [reply]
Yes, but I tried using the IP address in LAN, just to be safe.

**Comment by gAlleb** [comment, marked as answer]
Hi @Gvolten. If your mp3 file is generated on the same host I suggest a following scheme. It is better than using http server. I presume that you have daily reports.

1. Create a folder that will hold the generated weather.mp3 file. e.g.  ```mkdir /var/azuracast/weather```
2. Create a cronjob that will replace weather.mp3 file with the specific one in the above mentioned folder (a few hours before playback). Probably it should call some kind of bash script which eventually will do something like `cp weather.mp3 /var/azuracast/weather/` in the end.
3. Create or edit `docker-compose.override.yml` -> `nano /var/azuracast/docker-compose.override.yml` and mount a folder into azuracast:

```ruby
services:
  web:
    volumes:
      - /var/azuracast/weather:/var/azuracast/stations/***YOUR-STATION***/media/weather
```

4.  Restart azuracast.  
```
cd /var/azuracast
docker compose down
docker compose up -d
```
5.  Go to `Media -> Music Files`. You should see your `weather` folder there.
6. Create a new playlist (e.g. `weather_everyday`) and point it to your `weather` folder (this would help to prevent removal of the file from the playlist upon rewrite). Edit your playlist. Make it `Advanced`. Save changes.
7. Go to `Broadcasting -> Edit Liquidsoap Configuration -> Third Box` and paste:
```ruby
# We reload the playlist manually to be sure it holds the newest generated file. 
# If you are going to start your weather reports, say, at 7:00, then reload playlist at 6:00 to have some space between.

thread.when(predicate.activates({ 6h00m }), fun() -> playlist_weather_everyday.reload())
```
Then do whatever you want. You can use `switch` operator with `fallback`:
```ruby
# introduce new "weather" variable and times for weather announcements: 7h00m or 9h00m or 11h00m or 13h00m and so on.

weather = fallback(
  track_sensitive=false,
  [
    switch(
      id="weather_switch",
      [(predicate.once({ 7h00m or 9h00m or 11h00m or 13h00m or 15h00m or 17h00m or 19h00m or 21h00m }), playlist_weather_everyday)]
    )
  ]
)

# When weather source becomes active skip current track, play weather report and continue with new track
radio = fallback.skip(weather, fallback=radio)
```
<hr>

Option 2 (using http server):

1. You don't have to create any playlists in Azuracast. Make sure that your `weather.mp3` file is in a remote playlist: `https://path/to/playlist_file.m3u` (`playlist_file.m3u` should contain your `weather.mp3` like `https://path/to/weather.mp3`).
2. Go to `Broadcasting -> Edit Liquidsoap Configuration -> Third Box` and paste:
```ruby

# Create a playlist and point it to your remote playlist.
playlist_weather_everyday = playlist("https://path/to/playlist_file.m3u")

# We reload the playlist manually to be sure it holds the newest generated file. 
# If you are going to start your weather reports, say, at 7:00, then reload playlist at 6:00 to have some space between.

thread.when(predicate.activates({ 6h00m }), fun() -> playlist_weather_everyday.reload())
```
3. Then do whatever you want. You can use `switch` operator with `fallback`:
```ruby
# introduce new "weather" variable and times for weather announcements: 7h00m or 9h00m or 11h00m or 13h00m and so on.

weather = fallback(
  track_sensitive=false,
  [
    switch(
      id="weather_switch",
      [(predicate.once({ 7h00m or 9h00m or 11h00m or 13h00m or 15h00m or 17h00m or 19h00m or 21h00m }), playlist_weather_everyday)]
    )
  ]
)

# When weather source becomes active skip current track, play weather report and continue with new track
radio = fallback.skip(weather, fallback=radio)
```

<hr>

Option 3:
Alternatively you can do it via Azuracast tools but I'm not sure about the outcome:

1. Create playlist and point it to your remote file or to a playlist with this file
<img width="500" alt="image" src="https://github.com/user-attachments/assets/ec815d0b-45ff-4527-a40a-1534b6e36e12">

2. Make a lot of schedules - I mean A LOT OF!)) With `loop once` option.
<img width="500" alt="image" src="https://github.com/user-attachments/assets/cd206a0a-3cc3-4f3e-b8d8-5294e747aee8">

3. 
<img width="500" alt="image" src="https://github.com/user-attachments/assets/d16133f3-e2f8-401c-b75f-2d0d4b0a3936">

Not sure how this option will go. Nevertheless I think you should reload it manually too (add 1 line to the 3rd box -> `thread.when(predicate.activates({ 6h00m }), fun() -> playlist_weather_everyday.reload())`). Not sure about it) Therefore I suggest using the first or second option.

**Reply by Gvolten** [reply]
Thanks for the help.
What I did in the end was, that I added the capability to download a .m3u playlist with the report. The report is always updated with the current info, so I run some code every hour to update it, and after it generates I also update the URL of the playlist in Azuracast via its API (The name of the report is the hash of its contents). Then I added a bunch of schedules to the playlist and set all of the necessary options and it now plays correctly.
